sdks/go/pkg/beam/io/fhirio/common.go

```diff
@@ -58,38 +62,34 @@ func extractBodyFrom(response *http.Response) (string, error) {
 	return string(bodyBytes), nil
 }
 
-func isBadStatusCode(status string) bool {
-	// 2XXs are successes, otherwise failure.
-	isMatch, err := regexp.MatchString("^2\\d{2}", status)
-	if err != nil {
-		return true
-	}
-	return !isMatch
+type fhirStoreClient interface {
+	readResource(resourcePath string) (*http.Response, error)
+	executeBundle(storePath string, bundle []byte) (*http.Response, error)
+	search(storePath, resourceType string, queries map[string]string, pageToken string) (*http.Response, error)
 }
 
-type fhirioFnCommon struct {
-	client                fhirStoreClient
-	resourcesErrorCount   beam.Counter
-	resourcesSuccessCount beam.Counter
-	latencyMs             beam.Distribution
+type fhirStoreClientImpl struct {
+	fhirService *healthcare.ProjectsLocationsDatasetsFhirStoresFhirService
 }
 
-func (fnc *fhirioFnCommon) setup(namespace string) {
-	if fnc.client == nil {
-		fnc.client = newFhirStoreClient()
-	}
-	fnc.resourcesErrorCount = beam.NewCounter(namespace, baseMetricPrefix+"resource_error_count")
-	fnc.resourcesSuccessCount = beam.NewCounter(namespace, baseMetricPrefix+"resource_success_count")
-	fnc.latencyMs = beam.NewDistribution(namespace, baseMetricPrefix+"latency_ms")
+func (c *fhirStoreClientImpl) readResource(resourcePath string) (*http.Response, error) {
+	return c.fhirService.Read(resourcePath).Do()
 }
 
-type fhirStoreClient interface {
-	readResource(resourcePath string) (*http.Response, error)
-	executeBundle(storePath string, bundle []byte) (*http.Response, error)
+func (c *fhirStoreClientImpl) executeBundle(storePath string, bundle []byte) (*http.Response, error) {
+	return c.fhirService.ExecuteBundle(storePath, bytes.NewReader(bundle)).Do()
 }
 
-type fhirStoreClientImpl struct {
-	fhirService *healthcare.ProjectsLocationsDatasetsFhirStoresFhirService
+func (c *fhirStoreClientImpl) search(storePath, resourceType string, queries map[string]string, pageToken string) (*http.Response, error) {
+	queryParams := make([]googleapi.CallOption, 0)
+	for key, value := range queries {
+		queryParams = append(queryParams, googleapi.QueryParameter(key, value))
+	}
+
+	if pageToken != "" {
+		queryParams = append(queryParams, googleapi.QueryParameter(pageTokenParameterKey, pageToken))
+	}
+	return c.fhirService.Search(storePath, &healthcare.SearchResourcesRequest{ResourceType: resourceType}).Do(queryParams...)
```

Comment:
Ah... did you get this pattern from an example text somewhere? (If you did, could you link me it?)

It's unintended behaviour to be able to specify the resource type like this, and this caused a large outage for us a couple months back. I fixed this in the Java implementation (https://github.com/apache/beam/pull/17315, see more details at b/227655182)

Could you fix this, if resourceType is empty then you use [.Search](https://pkg.go.dev/google.golang.org/api/healthcare/v1#ProjectsLocationsDatasetsFhirStoresFhirService.Search), otherwise use [.SearchType](https://pkg.go.dev/google.golang.org/api/healthcare/v1#ProjectsLocationsDatasetsFhirStoresFhirService.SearchType)

Reply:
Ah, I didn't notice that. Good catch